drv/fpga-server/src/main.rs

```diff
@@ -117,11 +179,45 @@ impl<Device: FpgaUserDesign> Drop for UserDesignLock<'_, Device> {
 }
 
 impl<'a, Device: Fpga<'a> + FpgaUserDesign> ServerImpl<'a, Device> {
+    fn check_lock_and_get_device(
+        &self,
+        caller: userlib::TaskId,
+        device_index: u8,
+    ) -> Result<&'a Device, FpgaError> {
+        let device_index = usize::from(device_index);
+
+        match &self.lock_holder {
+            Some(lock_state) => {
+                // The fact that we received this message _at all_ means
+                // that the sender matched our closed receive, but just
+                // in case we have a server logic bug, let's check.
+                assert!(lock_state.task == caller);
+
+                if lock_state.device_index == device_index {
+                    Ok(&self.devices[device_index])
+                } else {
+                    Err(FpgaError::BadDevice)
+                }
+            }
+            None => {
+                if device_index < self.devices.len() {
+                    Ok(&self.devices[device_index])
+                } else {
+                    Err(FpgaError::BadDevice)
+                }
```

Comment:
Nit:
```suggestion
                self.devices.get(device_index).ok_or(FpgaError::BadDevice),
```

Reply:
Thanks! I managed to make this function even shorter.